### ISSUE TITLE
[naming-convention] Ignore naming imposed by external packages

### ISSUE THREAD

**Issue by patheticcockroach** [typescript-eslint/typescript-eslint#2160]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "warn",
      {
        "selector": "default",
        "format": ["camelCase", "UPPER_CASE"]
      }
    ]
  }
}
```

```TS
import AWS from 'aws-sdk';
const params: AWS.S3.HeadObjectRequest = {
  Bucket: 'bucket',
  Key: 'file'
};
```

**Expected Result**
No warning: the names don't match my defined patterns, but there's nothing I can do about it as they are defined in the package I'm importing

**Actual Result**
I'm getting those warnings:
```
Property name Bucket must match one of the following formats: camelCase, UPPER_CASE
Property name Key must match one of the following formats: camelCase, UPPER_CASE
```

**Additional Info**
I didn't have that issue with the old, now removed, camelcase rule (maybe that was thanks to the "ignoreImport" falg?)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` |
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `12.16.3` |
| `npm`                              | `6.14.5` |

**Comment by bradzacher** [comment]
See discussion in #2147

**Comment by patheticcockroach** [comment]
Not sure if I should reply here or in the discussion you mention, but listing all the names, one by one, in a regex in order to ignore them is really not an acceptable solution. In my example, I only put 2 of those, but in a real project I have dozens of them...

**Comment by bradzacher** [comment]
There's no way for the rule to tell that a name is "required by the types provided by an external package".

To me it seems like you want to allow `PascalCase` naming convention for properties in your codebase, as that's a convention you are required to use.
If you don't want to do this, then the attached issue is the way to list the exceptions to your naming convention.

You can also use eslint disable comments to ignore the specific property definitions you want to be exceptions.

**Comment by patheticcockroach** [comment]
The thing is, I have another package that uses snake_case (thanks Stripe!), so I'd have to allow that as well... What surprised me is that it used to work (a bit) better with the previous camelcase setting (Stripe got caught in it, but not AWS), so I thought maybe this was a fixable regression.

But since it's not, I think I'll just remove the whole naming-convention thing. I already have a over a hundred disable/enable comments for that rule, and with all the new warnings it's getting a bit out of hand 🤔

Thanks for the clarification

**Comment by bbarry** [comment]
@bradzacher would it be possible to enforce naming-convention or a similar rule on types but not on values?

That way it could apply on the types you create in your project but not on those you are importing from external sources.

**Comment by bradzacher** [comment]
> What surprised me is that it used to work (a bit) better with the previous camelcase setting

This is because eslint's `camelcase` rule is actually terribly loose and is (in a way) misnamed; it doesn't technically enforce a `camelcase` naming convention!

Instead it enforces that names to not have underscores anywhere but at the start/end (with the exception of the allowing `UPPER_CASE`). [repl](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGNhbWVsY2FzZTogJ2Vycm9yJyAqL1xuXG5jb25zdCBGb29CYXIgPSAxO1xuY29uc3QgeCA9IHtcbiAgRm9vQmFyOiAxLFxufTtcblxuZnVuY3Rpb24gVVBQRVJBTExUSEVUSElOR1MoKSB7fVxuY29uc3QgTVlfQ09OU1RBTlQgPSAxO1xuXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

This means that implicitly it allows `PascalCase` names as well.

This is one other reason that I mention that you might want to just add `PascalCase` to your allowed formats, because it used to be allowed anyway.

I would consider adding exceptions to your config on a case-by-case basis. Depending on how things are defined, you might be able to auto-generate the `filter` regex.  
At the very least by using a `.eslintrc.js`, you can make it a bit easier to define your exception list:

```js
const ignoredPropNames = `^(${[
  'name_one',
  'NameTwo',
  ...
].join('|')})$`;
```

----

> would it be possible to enforce naming-convention or a similar rule on types but not on values?

It's technically feasible, yes, but it's not easy, and would require always needing type information in the rule.

You must use type information to determine if a property is a "freehand" key, or if it's defined in the type on the type annotation. You can't do it purely by inspecting the AST because of cases like:

```ts
type MyMap = Record<string, unknown>;
const x: MyMap = { This_Name_Should_Be_Checked: 1 };

type MyAny = any;
const y: MyAny = { This_Name_Should_Be_Checked: 1 };

type MyObj = { This_Name_Should_NOT_Be_Checked: 1 };
const z: MyObj = { This_Name_Should_NOT_Be_Checked: 1 };

type Weird = { This_Name_Should_NOT_Be_Checked: 1, [k: string]: number }
const weird: Weird = {
    This_Name_Should_NOT_Be_Checked: 1,
    But_This_Name_Should_Be_Checked: 1,
};
```

The rule currently only requires type information if and only if you use the `types` selector. Requiring type information in more cases is a breaking change.
